pkg/sql/plan2/function/builtin/multi/date_add.go

```diff
@@ -25,126 +25,104 @@ import (
 
 func DateAdd(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
 	firstVector := vectors[0]
+	secondVector := vectors[1]
 	firstValues := vector.MustTCols[types.Date](vectors[0])
 	secondValues := vector.MustTCols[int64](vectors[1])
 	thirdValues := vector.MustTCols[int64](vectors[2])
 
 	resultType := types.Type{Oid: types.T_date, Size: 4}
 	resultElementSize := int(resultType.Size)
-	if firstVector.IsScalar() {
-		if firstVector.ConstVectorIsNull() {
-			return proc.AllocScalarNullVector(resultType), nil
-		}
-		resultVector := vector.NewConst(resultType)
-		resultValues := make([]types.Date, 1)
-		res, err := date_add.DateAdd(firstValues, secondValues, thirdValues, resultVector.Nsp, resultValues)
-		vector.SetCol(resultVector, res)
-		return resultVector, err
+	var maxLen int
+	if len(firstValues) > len(secondValues) {
+		maxLen = len(firstValues)
 	} else {
-		resultVector, err := proc.AllocVector(resultType, int64(resultElementSize*len(firstValues)))
-		if err != nil {
-			return nil, err
-		}
-		resultValues := encoding.DecodeDateSlice(resultVector.Data)
-		resultValues = resultValues[:len(firstValues)]
-		nulls.Set(resultVector.Nsp, firstVector.Nsp)
-		res, err := date_add.DateAdd(firstValues, secondValues, thirdValues, resultVector.Nsp, resultValues)
-		vector.SetCol(resultVector, res)
-		return resultVector, err
+		maxLen = len(secondValues)
 	}
+	resultVector, err := proc.AllocVector(resultType, int64(resultElementSize*maxLen))
+	if err != nil {
+		return nil, err
+	}
+	resultValues := encoding.DecodeDateSlice(resultVector.Data)
+	resultValues = resultValues[:maxLen]
+	res, err := date_add.DateAdd(firstValues, secondValues, thirdValues, firstVector.Nsp, secondVector.Nsp, resultVector.Nsp, resultValues)
+	vector.SetCol(resultVector, res)
+	return resultVector, err
 }
 
 func DatetimeAdd(vectors []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
 	firstVector := vectors[0]
+	secondVector := vectors[1]
 	firstValues := vector.MustTCols[types.Datetime](vectors[0])
 	secondValues := vector.MustTCols[int64](vectors[1])
 	thirdValues := vector.MustTCols[int64](vectors[2])
 
 	resultType := types.Type{Oid: types.T_datetime, Precision: firstVector.Typ.Precision, Size: 8}
 	resultElementSize := int(resultType.Size)
-	if firstVector.IsScalar() {
-		if firstVector.ConstVectorIsNull() {
-			return proc.AllocScalarNullVector(resultType), nil
-		}
-		resultVector := vector.NewConst(resultType)
-		resultValues := make([]types.Datetime, 1)
-		res, err := date_add.DatetimeAdd(firstValues, secondValues, thirdValues, resultVector.Nsp, resultValues)
-		vector.SetCol(resultVector, res)
-		return resultVector, err
+	var maxLen int
+	if len(firstValues) > len(secondValues) {
```

Comment:
Are you sure this is the right way to handle vector length?   Aren't they ALWAYS having matching length?